nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,227 @@
+			"version": {
+				Type:        schema.TypeString,
+				Description: "Version for the upgrade",
+				Required:    true,
+			},
+			"accept_user_agreement": {
+				Type:        schema.TypeBool,
+				Description: "Whether to accept the user agreement",
+				Required:    true,
+			},
+			"failed_prechecks": {
+				Type:        schema.TypeList,
+				Description: "List of failed prechecks for the upgrade, only include warnings, if precheck failed with error then the resource creation will fail",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Description: "ID of failed precheck",
+							Computed:    true,
+						},
+						"message": {
+							Type:        schema.TypeString,
+							Description: "Message of the failed precheck",
+							Computed:    true,
+						},
+						// "needs_ack": {
+						// 	Type:        schema.TypeBool,
+						// 	Description: "Flag which identifies if acknowledgement is required for the precheck",
+						// 	Computed:    true,
+						// },
+						// "needs_resolve": {
+						// 	Type:         schema.TypeBool,
+						// 	Description:  "Flag which identifies if resolution is required for the precheck",
+						// 	ValidateFunc: validation.IsIPAddress,
+						// 	Required:     true,
+						// },
+					},
+				},
+				Computed: true,
+			},
+		},
+	}
+}
+
+func uploadUpgradeBundle(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	bundleURL := d.Get("bundle_url").(string)
+	version := d.Get("version").(string)
+	client := upgrade.NewBundlesClient(connector)
+	installParam := false
+	bundleType := nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_PRE_UPGRADE
+	c := m.(nsxtClients)
+	userName := c.NsxtClientConfig.UserName
+	password := c.NsxtClientConfig.Password
+	upgradeBundleFetchRequest := nsxModel.UpgradeBundleFetchRequest{
+		BundleType: &bundleType,
+		Password:   &password,
+		Username:   &userName,
+		Url:        &bundleURL,
+		Version:    &version,
+	}
+	bundleID, err := client.Create(upgradeBundleFetchRequest, &installParam)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", "", err)
+	}
+	err = waitForBundleUpload(m, *bundleID.BundleId)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", "", err)
+	}
+	return nil
+}
+
+func acceptUserAgreement(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	acceptUserAgreement := d.Get("accept_user_agreement").(bool)
+	if !acceptUserAgreement {
+		return fmt.Errorf("To proceed with upgrade, you must accept user agreement")
+	}
+	client := eula.NewAcceptClient(connector)
+	err := client.Create()
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", "", err)
+	}
+	return nil
+}
+
+func executePreupgradeChecks(m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewUpgradeClient(connector)
+	return client.Executepreupgradechecks(nil, nil, nil, nil, nil, nil)
```

Comment:
I believe we should use a separate polling API to wait for the checks to complete

Reply:
An update on this: yes we do need a separate polling API, otherwise the failed prechecks that we retrieved from FailureClient will only be the ones that are already executed.